### ISSUE TITLE
-o - doesn't work

### ISSUE THREAD

**Issue by gi0e5b06** [LMMS/lmms#3890]
`-r project.mmp -o -` should send the output to stdout.
Now it writes a file named `-.wav`

**Comment by PhysSong** [comment]
>Now it writes a file named -.wav

It is correct according to LMMS manpage. AFAIK there's no reason `-o -` should mean "output to stdout". For example, `gcc` generates a file named `-` if `-o -` is given.
If one needs render to standard output, adding `--stdout` can be an solution.

**Comment by tresf** [comment]
> If one needs render to standard output, adding --stdout can be an solution.

Agreed.  `--render|-r`  outputs a single file whereas `--rendertracks` outputs several files.  I assume we'd want to support both.  I also don't have a problem with keeping `-o -` as @gi0e5b06 suggests.  @gi0e5b06 can you provide us a few command line applications that use this `-` flag for enabling `stdout`?

**Comment by tresf** [comment]
> can you provide us a few command line applications that use this `-` flag for enabling stdout?

Found it... https://stackoverflow.com/a/3797806/3196753.

**Comment by tresf** [comment]
This looks like it will be some refactoring... 

* `AudioFileDevice` interface [assumes we pass in a `QString` pointing to the output file](https://github.com/LMMS/lmms/blob/08201c202494c25ab8a151c98118dfa0359500e6/include/AudioFileDevice.h#L39) and the actual underlying device is doing the IO.  This means we'd need a special constructor to bypass physical file writing or use redundant logic in each class.
* We're [already writing to `stdout` as part of the rendering process](https://github.com/LMMS/lmms/blob/e9a406311969397a9d4219eb80671d27db826e87/src/core/main.cpp#L754).  We'd need to supress that when `-` is detected or politely and irresponsibly move it to `stderr`.

**Comment by gi0e5b06** [comment]
`./lmms -r ~/lmms/projects/test6.mmp -f au -o - | play -q -t au -`

Done. At least for a very first version.
That brings a lot of opportunities (streaming for example) but also a lot of possibilities (ideas for future development). It's quite late here so more info tomorrow.

**Comment by gi0e5b06** [comment]
@PhysSong The issue was more about support for stdout. As for -o, it is used by most programs, including lmms, for specifying the output. The most common pattern is
`[input_file [...]] [-o output_file]`
Since it is already the pattern used by lmms, I suggest __not to change__ it.
As for `"-"`, it's the usual way to represent stdin and stdout and the easiest to remember.

@tresf
- QFile can open FileDescriptor (so stdout). More important, libsndfile is also able to do it, at least for encoders that don't require seek().
- I changed the printf/fprintf to qInfo/qWarning/qCritical/qFatal.
- Typing `"-"` in the FileChoose is porbably not a problem. The file would have an absolute path and an extension added so /home/user/path/-.wav . Anyway `"-"` has a special meaning only in the command line.

Finally, the commit is here:
https://github.com/gi0e5b06/lmms/commit/aed0d44d9f993a61bebd5bb2285fe163a93bba77
